x/stake/keeper/validator.go

```diff
@@ -246,13 +257,73 @@ func (k Keeper) UpdateValidator(ctx sdk.Context, validator types.Validator) type
 			bz := k.cdc.MustMarshalBinary(validator.ABCIValidator())
 			store.Set(GetTendermintUpdatesKey(validator.Owner), bz)
 		}
-
 	}
 
 	k.SetValidator(ctx, validator)
 	return validator
 }
 
+// updateCliffValidator determines if the current cliff validator needs to be
+// updated or swapped. If the provided affected validator is the current cliff
+// validator before it's power was increased, either the cliff power key will
+// be updated or if it's power is greater than the next bonded validator by
+// power, it'll be swapped.
+func (k Keeper) updateCliffValidator(ctx sdk.Context, affectedVal types.Validator) {
+	store := ctx.KVStore(k.storeKey)
+	pool := k.GetPool(ctx)
+	cliffAddr := sdk.AccAddress(k.GetCliffValidator(ctx))
+
+	oldCliffVal, found := k.GetValidator(ctx, cliffAddr)
+	if !found {
+		panic(fmt.Sprintf("cliff validator record not found for address: %v\n", cliffAddr))
+	}
+
+	// NOTE: We get the power via affectedVal since the store (by power key)
+	// has yet to be updated.
+	affectedValPower := affectedVal.GetPower()
+
+	// Create a validator iterator ranging from smallest to largest by power
+	// starting the current cliff validator's power.
+	start := GetValidatorsByPowerIndexKey(oldCliffVal, pool)
+	end := sdk.PrefixEndBytes(ValidatorsByPowerIndexKey)
+	iterator := store.Iterator(start, end)
+
+	offset := 0
+	var newCliffVal types.Validator
+
+	for ; iterator.Valid() && offset < 1; iterator.Next() {
+		ownerAddr := iterator.Value()
+
+		currVal, found := k.GetValidator(ctx, ownerAddr)
+		if !found {
+			panic(fmt.Sprintf("validator record not found for address: %v\n", ownerAddr))
+		}
+
+		if currVal.Status != sdk.Bonded || currVal.Revoked {
+			panic(fmt.Sprintf("unexpected revoked or unbonded validator for address: %s\n", ownerAddr))
+		}
+
+		newCliffVal = currVal
+		offset++
+	}
+
+	iterator.Close()
+	if offset == 0 {
+		panic("unable to find cliff validator")
+	}
+
+	if bytes.Equal(affectedVal.Owner, newCliffVal.Owner) {
+		// The affected validator remains the cliff validator, however, since
+		// the store does not contain the new power, set the new cliff
+		// validator to the affected validator.
+		k.setCliffValidator(ctx, affectedVal, pool)
```

Comment:
We just need to update the power, right? I think `setCliffValidator` will unnecessarily re-set the cliff validator address.

Reply:
Correct! I suppose we can save a write, huh?